be/src/olap/cumulative_compaction_time_series_policy.cpp

```diff
@@ -0,0 +1,255 @@
+#include "olap/tablet.h"
+#include "util/time.h"
+namespace doris {
+
+TimeSeriesCumulativeCompactionPolicy::TimeSeriesCumulativeCompactionPolicy(
+        int64_t compaction_goal_size, int64_t compaction_file_count_threshold,
+        int64_t compaction_time_threshold_seconds)
+        : _compaction_goal_size(compaction_goal_size),
+          _compaction_file_count_threshold(compaction_file_count_threshold),
+          _compaction_time_threshold_seconds(compaction_time_threshold_seconds) {}
+
+uint32_t TimeSeriesCumulativeCompactionPolicy::calc_cumulative_compaction_score(Tablet* tablet) {
+    uint32_t score = 0;
+    bool base_rowset_exist = false;
+    const int64_t point = tablet->cumulative_layer_point();
+
+    int64_t total_size = 0;
+    RowsetMetaSharedPtr first_meta;
+    int64_t first_version = INT64_MAX;
+    // NOTE: tablet._meta_lock is hold
+    auto& rs_metas = tablet->tablet_meta()->all_rs_metas();
+    // check the base rowset and collect the rowsets of cumulative part
+    for (auto& rs_meta : rs_metas) {
+        if (rs_meta->start_version() < first_version) {
+            first_version = rs_meta->start_version();
+            first_meta = rs_meta;
+        }
+        // check base rowset
+        if (rs_meta->start_version() == 0) {
+            base_rowset_exist = true;
+        }
+        if (rs_meta->end_version() < point || !rs_meta->is_local()) {
+            // all_rs_metas() is not sorted, so we use _continue_ other than _break_ here.
+            continue;
+        } else {
+            // collect the rowsets of cumulative part
+            total_size += rs_meta->total_disk_size();
+            score += rs_meta->get_compaction_score();
+        }
+    }
+
+    if (first_meta == nullptr) {
+        return 0;
+    }
+
+    // If base version does not exist, but its state is RUNNING.
+    // It is abnormal, do not select it and set *score = 0
+    if (!base_rowset_exist && tablet->tablet_state() == TABLET_RUNNING) {
+        LOG(WARNING) << "tablet state is running but have no base version";
+        return 0;
+    }
+
+    // Condition 1: the size of input files for compaction meets the requirement of parameter _compaction_goal_size
+    if (total_size >= _compaction_goal_size) {
+        return score;
+    }
+
+    // Condition 2: the number of input files reaches the threshold specified by parameter _compaction_file_count_threshold
+    if (score >= _compaction_file_count_threshold) {
+        return score;
+    }
+
+    int64_t now = UnixMillis();
+    int64_t last_cumu = tablet->last_cumu_compaction_success_time();
+    if (last_cumu != 0) {
+        int64_t cumu_interval = now - last_cumu;
+
+        // Condition 3: the time interval between compactions exceeds the value specified by parameter _compaction_time_threshold_second
+        if (cumu_interval > _compaction_time_threshold_seconds * 1000) {
+            return score;
+        }
+    } else {
+        tablet->set_last_cumu_compaction_success_time(now);
```

Comment:
It seems to be wrong to update success time here, since it's not success.

Reply:
I have deleted the content, the system continues to function properly.